src/update_agent/mod.rs

```diff
@@ -433,4 +584,73 @@ mod tests {
+            checksum: "ostree-checksum".to_string(),
+            age_index: None,
+        };
+        let mut machine = UpdateAgentState::UpdateAvailable((update.clone(), 0));
+
+        machine.update_staged(update.clone());
+        assert_eq!(
+            machine,
+            UpdateAgentState::UpdateStaged((update.clone(), MAX_FINALIZE_POSTPONEMENTS))
+        );
+
+        // Set up empty interactive sessions.
+        let no_interactive_sessions: Vec<InteractiveSession> = vec![];
+        let can_finalize = machine.handle_interactive_sessions(&no_interactive_sessions);
+        assert!(can_finalize);
+        assert_eq!(
+            machine,
+            UpdateAgentState::UpdateStaged((update.clone(), MAX_FINALIZE_POSTPONEMENTS))
+        );
+
+        // Set up dummy interactive sessions.
+        let fake_tty_path = tempfile::tempdir_in("/tmp").unwrap();
+        let fake_tty_path_str = fake_tty_path.path().to_str().unwrap();
+        let fake_tty = format!("{}/tty1", fake_tty_path_str);
+        let fake_session = InteractiveSession {
+            user: String::from("fakeuser"),
+            tty_dev: String::from(&fake_tty),
+        };
+        let interactive_sessions_present: Vec<InteractiveSession> = vec![fake_session];
+
+        // Postpone MAX_FINALIZE_POSTPONEMENTS times (counting from 1).
+        for finalization_attempt in 1..MAX_FINALIZE_POSTPONEMENTS + 1 {
+            let can_finalize = machine.handle_interactive_sessions(&interactive_sessions_present);
+            assert!(!can_finalize);
+            machine.record_postponement(); // as we cannot finalize.
+            let postponement_remaining =
+                MAX_FINALIZE_POSTPONEMENTS.saturating_sub(finalization_attempt);
+            assert_eq!(
+                machine,
+                UpdateAgentState::UpdateStaged((update.clone(), postponement_remaining))
+            )
+        }
+        // Sanity check final broadcasted message.
+        let tty_contents = fs::read_to_string(&fake_tty).unwrap();
+        assert!(tty_contents.contains("Broadcast message from Zincati"));
+        assert!(tty_contents.contains(&update.version));
+        assert!(tty_contents.contains(&format_seconds(DEFAULT_POSTPONEMENT_TIME_SECS)));
+
+        // Maximum allowed postponements reached.
+        let can_finalize = machine.handle_interactive_sessions(&interactive_sessions_present);
+        assert!(can_finalize);
+        assert_eq!(machine, UpdateAgentState::UpdateStaged((update.clone(), 0)));
+    }
+
+    #[test]
+    fn test_format_seconds() {
+        assert_eq!("1 second", format_seconds(1));
+        assert_eq!("2 seconds", format_seconds(2));
+        assert_eq!("1 minute", format_seconds(60));
+        assert_eq!("1 minute and 1 second", format_seconds(61));
+        assert_eq!("1 minute and 30 seconds", format_seconds(90));
+        assert_eq!("2 minutes", format_seconds(120));
+        assert_eq!("42 minutes and 23 seconds", format_seconds(2543));
```

Comment:
It's easier to believe these are correctly written if you write e.g. 2543 as `42*60 + 23`. :)

Reply:
haha, right, just `2543` really isn't too useful there :)